// next.js/lib/next.js
(function(factory) {

// node
if (typeof require !== 'undefined') {
  factory(require, exports, module);
  return;
}

// seajs
if (typeof define !== 'undefined') {
  define(factory);
  return;
}

// window
var _module = {
  exports: {}
};

factory(null, _module.exports, _module);

window['next'] = _module.exports;

})(function(require, exports, module) {
  
// ECMAScript5 is required 

var slice = [].slice;
var arrayForEach = [].forEach;
var arrayMap = [].map;
var noop = function() {};

var detachCallback = function(fn) {
  return function() {
    var a = arguments;
    return fn(slice.call(a, 0, a.length - 1), a[a.length - 1]);
  }
};

var dissect = function(handleArgs, handleResults) {
  return detachCallback(function(args, callback) {
    handleArgs.apply(null, args.concat([function(err) {
      handleResults(callback, err, arguments, args);
    }]));
  });
};

var extract = function(args, callback) {
  callback.apply(null, [null].concat(args));
};

var map0 = function(items, fn, callback) {
  var len = items.length;
  var results = [];

  if (!items.length) {
    callback(null, []);
    return;
  }

  var check = function(index, args) {
    if (!results[index]) {
      len -= 1;
      results[index] = args;
      if (len === 0) {    
        callback(null, results.reduce(function(ret, items) {
          return ret.concat(items);
        }));
        results = null;
        check = noop;
      }
    }
  };

  var isError = false;
  arrayForEach.call(items, function(item, i) {
    fn(item, function(err) {
      if (isError) {
        return;
      }
      if (err) {
        callback(err);
        isError = true;
        return;
      }
      check(i, slice.call(arguments, 1));
    })
  });
};

var normalizeCallback = function(fn) {
  return typeof fn === 'function' ? fn : detachCallback(function(args, callback) {
    callback(null, fn);
  });
};

var normalizeCallbacks = function(fns) {
  return arrayMap.call(fns, normalizeCallback);
};

var reduceArgs = function(fn) {
  return function() {
    return arguments.length ? normalizeCallbacks(arguments).reduce(fn) : echo;
  }
};

var echo = detachCallback(extract);

var pipe = reduceArgs(function(prev, next) {
  return dissect(prev, function(callback, err, args) {
    if (err) {
      callback(err);
    } else {
      next.apply(null, slice.call(args, 1).concat([callback]));
    }
  });
});

var attempt = reduceArgs(function(prev, next) {
  return dissect(prev, function(callback, err, args, lastArgs) {
    if (err) {
      next.apply(null, lastArgs.concat([callback]));
    } else {
      callback.apply(null, args);
    }
  });
});

var map = function(fn) {
  return detachCallback(function(args, callback) {
    var items = args[0];
    var extraArgs = args.slice(1);
    return map0(items, extraArgs.length ? detachCallback(function(itemArgs, cb) {
      fn.apply(null, itemArgs.concat(extraArgs, [cb]));
    }) : fn, callback);
  })
};

var parallel = function() {
  var fns = normalizeCallbacks(arguments);
  return pipe(
    detachCallback(function(args, callback) {
      map0(fns, function(fn, cb) {
        fn.apply(null, args.concat([cb]));
      }, callback);
    }),
    extract
  );
};

var concurrency = function(fn, limit, onDrain) {
  var buffer = [];
  var runnings = 0;

  var allocate = function() {
    if (runnings !== limit) {
      var args = buffer.shift();
      if (args) {
        runnings += 1;
        run(args);
      } else {        
        if (!runnings) {
          if (onDrain) {
            onDrain();
          }
        }
      }
    }
  };

  var release = function() {
    runnings -= 1;
    allocate();
  };

  var run = function(a) {
    var callback = a[a.length - 1];
    return fn.apply(null, slice.call(a, 0, a.length - 1).concat([function() {
      release();
      callback.apply(null, arguments);
    }]));
  };

  return function() {
    buffer.push(arguments);
    allocate();
  };
};

var genKeyFromFirstArg = function(x) {
  return x;
};

var memoize = function(fn, genKeyFromArgs) {
  if (!genKeyFromArgs) {
    genKeyFromArgs = genKeyFromFirstArg;
  }

  var cache = {};
  var buffer = {};

  return detachCallback(function(args, callback) {
    var key = genKeyFromArgs.apply(null, args);
    if (cache[key]) {
      callback.apply(null, cache[key]);
      return;
    }

    if (buffer[key]) {
      buffer[key].push(callback);
      return;
    }

    buffer[key] = [callback];
    fn.apply(null, args.concat([function() {
      var args = cache[key] = arguments;
      buffer[key].forEach(function(cb) {
        cb.apply(null, args);
      });
      delete buffer[key];
    }]));
  });
};

module.exports = {
  pipe: pipe,
  attempt: attempt,
  map: map,
  parallel: parallel,
  concurrency: concurrency,
  memoize: memoize,
  callback: {
    map: map0,
    extract: extract,
    echo: echo
  }
};

});
